src/libponyc/verify/fun.c

```diff
@@ -487,6 +487,95 @@ static bool show_partiality(pass_opt_t* opt, ast_t* ast)
   return false;
 }
 
+// From https://github.com/ponylang/ponyc/issues/3765:
+//
+// When an immutable object is sent across actors, the garbage collector traces
+// its associated object graph, using the associated type descriptor.
+// Structs don't have a type descriptor, so the GC can't trace them.
+//
+// An option is to track additional information for objects without type
+// descriptors, but this incurs in severe overhead for the normal case.
+//
+// Another option is to disallow non-opaque objects without type descriptors to
+// be the root of an object graph. Essentially, this means that one has to wrap
+// non-tag structs (and tuple containing non-tag structs, since the GC will try
+// to trace each element) inside another object with a type descriptor.
+bool verify_behaviour_root_struct_param(pass_opt_t* opt, ast_t* type, bool is_in_tuple)
+{
+  // For correct error location, keep track of the original AST node,
+  // even if we have to use the node's data as type node.
+  ast_t* type_node = type;
+  ast_t* cap = NULL;
+  if (ast_id(type) == TK_NOMINAL)
+  {
+    // We're still interested in the capability of the type as it was used,
+    // so save that information.
+    cap = ast_childidx(type, 3);
+    type = (ast_t*)ast_data(type);
+  }
+
+  switch(ast_id(type))
+  {
+    case TK_STRUCT:
+    {
+      if(cap == NULL)
+        cap = ast_childidx(type, 2);
+
+      if(ast_id(cap) == TK_TAG)
+        return true;
+
+      if(is_in_tuple)
+      {
+        ast_error(opt->check.errors, type_node,
+          "cannot use a tuple with a struct type member as parameter to a behaviour; "
+          "this would be possible if the capability was \"tag\", "
+          "or if the parameter was wrapped inside a class or actor");
```

Comment:
I think something along the lines of:

"only "tag"\ structs can be used as a parameter to a behavior; "
"you'll need to change the capability or wrap the struct in a class or actor"

would be more clear. although i also think that could be improved.